### PR TITLE
Develop fbph

### PR DESCRIPTION
**Pull request by mikewin-climsci**
Implementation of biophysical (bph) effects of afforestation. Include remote sensing data of bph into the carbon prize driven afforestation benefit in 32_forestry and 56_ghg_policy.

### CODE REVIEW

**Comment by mikewin-climsci** [comment]
> 
> 
> Seems all good to me.
> Has the model performance been evaluated? If yes then this can be merged

Is there an agreed-upon performance testing scheme that I can follow? Current default runs (revision 15.05. 22:49) clock in at 31.8mins. Default runs with my commits (rev. 13.05 17:06!) took 31.2mins.

**Comment by abhimishr** [comment]
> 
> 
> > Seems all good to me.
> > Has the model performance been evaluated? If yes then this can be merged
> 
> Is there an agreed-upon performance testing scheme that I can follow? Current default runs (revision 15.05. 22:49) clock in at 31.8mins. Default runs with my commits (rev. 13.05 17:06!) took 31.2mins.

Current default's runtime and overall model behaviour should be the benchmark. Also nobgp will be the default so it looks okay I guess.

File: modules/32_forestry/dynamic_oct19/sets.gms [hunk 1]
```diff
@@ -19,5 +19,11 @@ sets
  kforestry(kall) Forestry products
  / wood, woodfuel /
 
+ bgp32 beschreibung
+ / nobgp, ann, djf, jja /
+
+ aff_effect Beschreibung
```
Comment (flohump) [inline modules/32_forestry/dynamic_oct19/sets.gms:25]: Replace "Beschreibung" with proper description

File: modules/32_forestry/dynamic_oct19/sets.gms [hunk 2]
```diff
@@ -19,5 +19,11 @@ sets
  kforestry(kall) Forestry products
  / wood, woodfuel /
 
+ bgp32 beschreibung
```
Comment (flohump) [inline modules/32_forestry/dynamic_oct19/sets.gms:22]: Replace "Beschreibung" with proper description

File: modules/56_ghg_policy/price_jan20/declarations.gms [hunk 3]
```diff
@@ -47,6 +47,8 @@ variables
  v56_emission_costs_reg_oneoff(i,emis_reg_one56)         Costs for emissions occuring only once in time (mio. USD05MER per yr)
  v56_emission_costs_cell_yearly(j,emis_cell_yr56)        Costs for emissions occuring yearly (mio. USD05MER per yr)
  v56_emission_costs_cell_oneoff(j,emis_cell_one56)       Costs for emissions occuring only once in time (mio. USD05MER per yr)
+ vm_reward_cdr_aff(i)                                    Regional revenues for carbon captured by afforestation (mio. USD05MER per yr)
```
Comment (flohump) [inline modules/56_ghg_policy/price_jan20/declarations.gms:50]: regional average annual expected revenue from afforestation

File: modules/56_ghg_policy/price_jan20/declarations.gms [hunk 4]
```diff
@@ -47,6 +47,8 @@ variables
...
+ v56_reward_cdr_aff(j)				                           Cellular revenues for carbon captured by afforestation (mio. USD05MER per yr)
```
Comment (flohump) [inline modules/56_ghg_policy/price_jan20/declarations.gms:51]: cellular average annual expected revenue from afforestation

File: modules/56_ghg_policy/price_jan20/equations.gms [hunk 5]
```diff
@@ -96,7 +96,7 @@ q56_cell_to_reg(i2,pollutants,emis_source) ..
                  ;
 
 *' The value of CDR from C-price induced afforestation enters the objective function as negative costs.
-*' The reward, which serves as incentive for afforestation, is calculated in 3 steps: 
+*' The reward, which serves as incentive for afforestation, is calculated in 3 steps:
 *' First, the expected CDR for each 5-year age-class and the corresponding future C price are multiplied.
```
Comment (flohump) [inline modules/56_ghg_policy/price_jan20/equations.gms:100]: Add here that now also bgp effects are included, next to bgc effects (CDR).

File: modules/32_forestry/dynamic_oct19/equations.gms [hunk 6]
```diff
@@ -25,9 +25,13 @@ vm_cost_fore(i2) =e= sum((cell(i2,j2),type32,ac,fcosts32),
 *' activity for a given planning horizon `s32_planing_horizon` to the [56_ghg_policy] module.
```
Comment (flohump) [inline modules/32_forestry/dynamic_oct19/equations.gms:25]: The interface `vm_cdr_aff` provides the projected bgc and bph effects of an afforestation activity...

File: modules/32_forestry/dynamic_oct19/declarations.gms [hunk 7]
```diff
@@ -27,7 +28,10 @@ positive variables
  vm_landdiff_forestry                    Aggregated difference in forestry land compared to previous timestep (mio. ha)
  v32_land_expansion(j,type32,ac) 		 Forestry land expansion compared to previous timestep (mio. ha)
  v32_land_reduction(j,type32,ac) 		 Forestry land reduction compared to previous timestep (mio. ha)
- vm_cdr_aff(j,ac) 						 Expected CDR from afforestation depending on planning horizon (mio. tC)
+;
+
+variables
+vm_cdr_aff(j,ac,aff_effect) 						 Expected CDR from afforestation depending on planning horizon (mio. tC)
```
Comment (flohump) [inline modules/32_forestry/dynamic_oct19/declarations.gms:34]: Exptected bgc and bph effects of afforestation ...

File: config/default.cfg [hunk 8]
```diff
@@ -416,6 +420,19 @@ cfg$gms$s32_max_aff_area <- Inf              # def = Inf
 # * ("onlytropical"): Afforestation only in tropical areas 20deg S-20deg N
 cfg$gms$c32_aff_mask <- "noboreal"            # def = "noboreal"
 
+# Switch to determine whether biogeophysical responses to
```
Comment (flohump) [inline config/default.cfg:423]: Technically, c32_aff_mask and c32_aff_bgp can be combined.
However, c32_aff_mask is exogenous, while c32_aff_bgp is endogenous .
Therefore, I can't think of a meaninful application that uses both switches at the same time. 
Maybe you can add this hint.

**Comment by flohump** [comment]
Please also add a line in the changelog file

**Comment by mikewin-climsci** [comment]
> 
> 
> The implementation looks fine to me. I have some suggestions for improving the definition of variables (see my comments). Please also mention bph effects in the corresponding realization.gms files of 56_ghg_policy and 32_foresty.

Thank you for the thorough check! I committed the requested descriptions in both modules, the default config, and the changelog.

**Comment by tscheypidi** [comment]
looks good to me. No additional requests from my side.

**Comment by abhimishr** [comment]
Seems like everybody is happy. I'll merge this!

**Comment by abhimishr** [comment]
Set description needed a small bugfix. PR https://github.com/magpiemodel/magpie/pull/185 should be able to fix that

**Pull request merged**